File: android/beagle/src/main/java/br/com/zup/beagle/android/action/Navigate.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.action

import android.view.View
import br.com.zup.beagle.android.annotation.ContextDataValue
import br.com.zup.beagle.android.components.layout.Screen
import br.com.zup.beagle.android.context.Bind
import br.com.zup.beagle.android.context.expressionOrValueOf
import br.com.zup.beagle.android.context.normalizeContextValue
import br.com.zup.beagle.android.networking.HttpMethod
import br.com.zup.beagle.android.utils.evaluateExpression
import br.com.zup.beagle.android.view.custom.BeagleNavigator
import br.com.zup.beagle.android.widget.RootView
import br.com.zup.beagle.core.BeagleJson
import br.com.zup.beagle.newanalytics.ActionAnalyticsConfig

/**
 * Class handles transition actions between screens in the application. Its structure is the following:.
 */

@BeagleJson(name = "navigate")
sealed class Navigate : AnalyticsAction {

    /**
     * Opens one of the browsers available on the device with the passed url.
     * @param url defined route to be shown.
     */
    @BeagleJson(name = "openExternalURL")
    data class OpenExternalURL(
        val url: Bind<String>,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {

        constructor(
            url: String,
            analytics: ActionAnalyticsConfig? = null,
        ) : this(
            url = expressionOrValueOf(url),
            analytics = analytics
        )

        override fun execute(rootView: RootView, origin: View) {
            evaluateExpression(rootView, origin, url)?.let { url ->
                BeagleNavigator.openExternalURL(rootView.getContext(), url)
            }
        }
    }

    /**
     * This action opens the route to execute the action declared in the deeplink that was defined for the application.
     * @param route deeplink identifier
     * @param shouldResetApplication attribute that allows customization of the behavior of
     * restarting the application view stack.
     * @param data pass information between screens.
     */
    @BeagleJson(name = "openNativeRoute")
    class OpenNativeRoute(
        val route: Bind<String>,
        val shouldResetApplication: Boolean = false,
        val data: Map<String, String>? = null,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {

        constructor(
            route: String,
            shouldResetApplication: Boolean = false,
            data: Map<String, String>? = null,
            analytics: ActionAnalyticsConfig? = null,
        ) : this(
            route = expressionOrValueOf(route),
            shouldResetApplication = shouldResetApplication,
            data = data,
            analytics = analytics
        )

        override fun execute(rootView: RootView, origin: View) {
            evaluateExpression(rootView, origin, route)?.let { route ->
                BeagleNavigator.openNativeRoute(rootView, route, data, shouldResetApplication)
            }
        }
    }

    /**
     * This action closes the current view stack.
     */
    @BeagleJson(name = "popStack")
    class PopStack(override var analytics: ActionAnalyticsConfig? = null) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.popStack(rootView.getContext())
        }
    }

    /**
     * Action that closes the current view.
     */
    @BeagleJson(name = "popView")
    class PopView(override var analytics: ActionAnalyticsConfig? = null) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.popView(rootView.getContext())
        }
    }

    /**
     * It is responsible for returning the stack of screens in the application flow to a specific screen.
     *
     * @param route route of a screen that it's on the pile.
     */
    @BeagleJson(name = "popToView")
    data class PopToView(
        val route: Bind<String>,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {

        constructor(
            route: String,
            analytics: ActionAnalyticsConfig? = null,
        ) : this(
            route = expressionOrValueOf(route),
            analytics = analytics
        )

        override fun execute(rootView: RootView, origin: View) {
            evaluateExpression(rootView, origin, route)?.let { route ->
                BeagleNavigator.popToView(rootView.getContext(), route)
            }
        }
    }

    /**
     * This type means the action to be performed is the opening
     * of a new screen using the route passed.
     * This screen will also be stacked at the top of the hierarchy of views in the application flow.
     *
     * @param route this defines navigation type, it can be a navigation to a remote route in which Beagle will
     * deserialize the content or to a local screen already built.
     */
    @BeagleJson(name = "pushView")
    data class PushView(val route: Route, override var analytics: ActionAnalyticsConfig? = null) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.pushView(rootView.getContext(), route.getSafe(rootView, origin))
        }
    }

    /**
     * Present a new screen with the link declared in the route attribute.
     * This attribute basically has the same functionality as PushView but starting a new flow instead.
     *
     * @param route this defines navigation type, it can be a navigation to a remote route in which Beagle will
     * deserialize the content or to a local screen already built.
     * @param controllerId in this field passes the id created in the custom activity for beagle to create the flow,
     * if not the beagle passes default activity.
     */
    @BeagleJson(name = "pushStack")
    data class PushStack(
        val route: Route,
        val controllerId: String? = null,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.pushStack(rootView.getContext(), route.getSafe(rootView, origin), controllerId)
        }
    }

    /**
     * This attribute, when selected, opens a screen with the route informed
     * from a new flow and clears clears the view stack for the entire application.
     *
     * @param route this defines navigation type, it can be a navigation to a remote route in which Beagle will
     * deserialize the content or to a local screen already built.
     * @param controllerId in this field passes the id created in the custom activity for beagle to create the flow,
     * if not the beagle passes default activity.
     */
    @BeagleJson(name = "resetApplication")
    data class ResetApplication(
        val route: Route,
        val controllerId: String? = null,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.resetApplication(rootView.getContext(), route.getSafe(rootView, origin), controllerId)
        }
    }

    /**
     * This attribute, when selected, opens a screen with the route informed
     * from a new flow and clears the stack of previously loaded screens.
     *
     * @param route this defines navigation type, it can be a navigation to a remote route in which Beagle will
     * deserialize the content or to a local screen already built.
     * @param controllerId in this field passes the id created in the custom activity for beagle to create the flow,
     * if not the beagle passes default activity.
     */
    @BeagleJson(name = "resetStack")
    data class ResetStack(
        val route: Route,
        val controllerId: String? = null,
        override var analytics: ActionAnalyticsConfig? = null,
    ) : Navigate() {
        override fun execute(rootView: RootView, origin: View) {
            BeagleNavigator.resetStack(rootView.getContext(), route.getSafe(rootView, origin), controllerId)
        }
    }

    internal fun Route.getSafe(rootView: RootView, origin: View): Route {
        if (this is Route.Remote) {
            val newValue = evaluateExpression(rootView, origin, url)
            val body = httpAdditionalData?.body?.normalizeContextValue()
                ?.let { evaluateExpression(rootView, origin, it) }

            return this.copy(
                url = Bind.Value(newValue ?: ""),
                httpAdditionalData = httpAdditionalData?.copy(body = body),
            )
        }
        return this
    }
}

/**
 * This defines navigation type,
 * it can be a navigation to a remote route in which Beagle will deserialize the content
 * or to a local screen already built.
 */
@BeagleJson
sealed class Route {
    /**
     * Class that takes care of navigation to remote content.
     * @param url attribute that contains the navigation endpoint.
     * @param shouldPrefetch tells Beagle if the navigation request should be previously loaded or not.
     * @param fallback screen that is rendered in case the request fails.
     * @param httpAdditionalData additional parameters to request
     */
    @BeagleJson
    data class Remote constructor(
        val url: Bind<String>,
        val shouldPrefetch: Boolean? = null,
        val fallback: Screen? = null,
        val httpAdditionalData: HttpAdditionalData? = null,
    ) : Route() {

        constructor(
            url: String,
            shouldPrefetch: Boolean? = null,
            fallback: Screen? = null,
            httpAdditionalData: HttpAdditionalData? = null,
        ) : this(
            expressionOrValueOf(url),
            shouldPrefetch,
            fallback,
            httpAdditionalData
        )
    }

    /**
     * Class indicating navigation to a local screen.
     * @param screen screen to be rendered.
     */
    @BeagleJson
    data class Local(
        val screen: Screen,
    ) : Route()
}

/**
 * HttpAdditionalData is used to do requests.
 * @param method HTTP method.
 * @param headers Header items for the request.
 * @param body Content that will be delivered with the request.
 */
@BeagleJson
data class HttpAdditionalData(
    val method: HttpMethod? = HttpMethod.GET,
    val headers: Map<String, String>? = hashMapOf(),
    @ContextDataValue
    val body: Any? = null,
)

File: iOS/Schema/Sources/Action/Types/Navigate/Navigate.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// Handles screens navigations actions of the application.
public enum Navigate: RawAction {
    
    /// Opens up an available browser on the device and navigates to a specified URL as String.
    case openExternalURL(String)
    
  /// Opens a screen that is defined completely local in your app (does not depend on Beagle) which will be retrieved using `DeeplinkScreenManager`.
    case openNativeRoute(OpenNativeRoute)

    /// Resets the application's root navigation stack with a new navigation stack that has `Route` as the first view
    case resetApplication(Route)
    
    /// Resets the views stack to create a new flow with the passed route.
    case resetStack(Route)
    
    /// Presents a new screen that comes from a specified route starting a new flow.
    /// You can specify a controllerId, describing the id of navigation controller used for the new flow.
    case pushStack(Route, controllerId: String? = nil)
    
    /// Unstacks the current view stack.
    case popStack

    /// Opens a new screen for the given route and stacks that at the top of the hierarchy.
    case pushView(Route)
    
    /// Dismisses the current view.
    case popView
    
    /// Returns the stack of screens in the application flow for a given screen in a route specified as String.
    case popToView(String)
    
    public struct OpenNativeRoute {
        
        /// Deeplink identifier.
        public let route: String
        
        /// Data that could be passed between screens.
        public let data: [String: String]?
        
        /// Allows customization of the behavior of restarting the application view stack.
        public let shouldResetApplication: Bool

        public init(
            route: String,
            data: [String: String]? = nil,
            shouldResetApplication: Bool = false
        ) {
            self.route = route
            self.data = data
            self.shouldResetApplication = shouldResetApplication
        }
    }
}

/// Defines a navigation route type which can be `remote` or `declarative`.
public enum Route {
    
    /// Navigates to a remote content on a specified path.
    case remote(NewPath)
    
    /// Navigates to a specified screen.
    case declarative(Screen)
}

extension Route {
    
    /// Constructs a new path to a remote screen.
    public struct NewPath {
        
        /// Contains the navigation endpoint.
        public let url: Expression<String>
        
        /// Changes _when_ this screen is requested.
        ///
        /// - If __false__, Beagle will only request this screen when the Navigate action gets triggered (e.g: user taps a button).
        /// - If __true__, Beagle will trigger the request as soon as it renders the component that have
        /// this action. (e.g: when a button appears on the screen it will trigger)
        public let shouldPrefetch: Bool
        
        /// A screen that should be rendered in case of request fail.
        public let fallback: Screen?

        /// Constructs a new path to a remote screen.
        ///
        /// - Parameters:
        ///   - url: Contains the navigation endpoint. Since its a _ExpressibleString_ type you can pass a Expression<String> or a regular String.
        ///   - shouldPrefetch: Changes _when_ this screen is requested.
        ///   - fallback: A screen that should be rendered in case of request fail.
        public init(url: StringOrExpression, shouldPrefetch: Bool = false, fallback: Screen? = nil) {
            self.url = "\(url)"
            self.shouldPrefetch = shouldPrefetch
            self.fallback = fallback
        }
    }
}

// MARK: Decodable

extension Navigate.OpenNativeRoute: Decodable {}

extension Navigate: Decodable {
    
    enum CodingKeys: CodingKey {
        case _beagleAction_
        case route
        case url
        case controllerId
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: ._beagleAction_)
        switch type.lowercased() {
        case "beagle:openexternalurl":
            self = .openExternalURL(try container.decode(String.self, forKey: .url))
        case "beagle:opennativeroute":
            self = .openNativeRoute(try .init(from: decoder))
        case "beagle:resetapplication":
            self = .resetApplication(try container.decode(Route.self, forKey: .route))
        case "beagle:resetstack":
            self = .resetStack(try container.decode(Route.self, forKey: .route))
        case "beagle:pushstack":
            self = .pushStack(try container.decode(Route.self, forKey: .route),
                              controllerId: try container.decodeIfPresent(String.self, forKey: .controllerId))
        case "beagle:popstack":
            self = .popStack
        case "beagle:pushview":
            self = .pushView(try container.decode(Route.self, forKey: .route))
        case "beagle:popview":
            self = .popView
        case "beagle:poptoview":
            self = .popToView(try container.decode(String.self, forKey: .route))
        default:
            throw DecodingError.dataCorruptedError(forKey: ._beagleAction_,
                                                   in: container,
                                                   debugDescription: "Can't decode '\(type)'")
        }
    }
}

extension Route: Decodable {
    
    enum CodingKeys: CodingKey {
        case type
        case path
        case shouldPrefetch
        case screen
        case data
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let screen = try? container.decode(ScreenComponent.self, forKey: .screen) {
            self = .declarative(screen.toScreen())
        } else {
            let newPath: Route.NewPath = try .init(from: decoder)
            self = .remote(newPath)
        }
    }
}

extension Route.NewPath: Decodable {
    
    enum CodingKeys: CodingKey {
        case url
        case shouldPrefetch
        case fallback
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.url = try container.decode(Expression<String>.self, forKey: .url)
        self.shouldPrefetch = try container.decodeIfPresent(Bool.self, forKey: .shouldPrefetch) ?? false
        self.fallback = try container.decodeIfPresent(ScreenComponent.self, forKey: .fallback)?.toScreen()
    }
}
